### PR TITLE
Add GoToTypeDefinition provider

### PR DESCRIPTION
**Pull request by jtschuster**
PR https://github.com/OmniSharp/omnisharp-roslyn/pull/2315 add a `/gototypedefinition` endpoint. This PR adds a VS Code provider that utilizes the endpoint.

Fixes https://github.com/OmniSharp/omnisharp-vscode/issues/4251

### CODE REVIEW

**Comment by 333fred** [comment]
I'll try to take a look at this over the next couple of nights.

File: src/features/typeDefinitionProvider.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        let req = <GoToTypeDefinitionRequest>createRequest(document, position);
+        req.WantMetadata = true;
+
+        const locations: Location[] = [];
+        try {
+            const goToTypeDefinitionResponse = await serverUtils.goToTypeDefinition(this._server, req, token);
+            // the defintion is in source
+            if (goToTypeDefinitionResponse && goToTypeDefinitionResponse.Definitions) {
+
+                for (const definition of goToTypeDefinitionResponse.Definitions) {
```
Comment (333fred) [inline src/features/typeDefinitionProvider.ts:36]: This logic should be sharable with https://github.com/OmniSharp/omnisharp-vscode/blob/090acf09c8/src/features/definitionProvider.ts#L36, pretty much in its entirety. Let's refactor these so they don't have to be maintained separately.

**Review by 333fred — CHANGES_REQUESTED** [review]

File: src/features/typeDefinitionProvider.ts (hunk 1 above)
Comment (jtschuster) [reply src/features/typeDefinitionProvider.ts]: I've updated to use the same `CSharpDefinitionProvider` class to implement both `DefinitionProvider` and `TypeDefinitionProvider`, and put as much as I could into a shared function.

**Review by 333fred — APPROVED** [review]
LGTM (commit 6). @JoeRobich for another review.

**Comment by jtschuster** [comment]
Forgot I was using a local build of Omnisharp with the new endpoint for local testing. The "/gototypedefinition" endpoint was just added recently, so this will fail in CI until it uses the next Omnisharp release that includes it.

**Comment by JoeRobich** [comment]
Thanks @jtschuster!

**Comment by crawlchange** [comment]
F12 is still not working. "Go to Type Definition" or "Go to Definition" are still not working. Does OmniSharp do ANYTHING at all?

**Pull request merged**